### ISSUE TITLE
[enhancement] two options to help bring back i3 3.x or wmii column behavior

### ISSUE THREAD

**Issue by quark-zju** [i3/i3#2100]
### Background

Initially I was a wmii user and switched to i3 for better xrandr and tray icon support.

After using i3 for years, I finally conclude the tree concept brings more trouble than benifits to me.

For example, consider the following layout:

```
  +------+------+
  |      |      |
  +------+------+
  |      |active|
  +------+------+
```

To create it, i3 3.x or wmii only needs 5 keystrokes. While i3 4.x requires 2 more (split v), which is 40%, a significant number.

Another annoying fact is it is unpredictable. Say I have the above layout created (in an unknown way) and before I press Mod+enter, I have no idea what the next state will be:

```
  +------+------+    +------+------+ 
  |      |      |    |      +------+ 
  +----+-+--+---+    +------+      | 
  |    |    |   |    |      +------+ 
  +----+----+---+    +------+------+ 
```

Basically it's like in browser, `<div><div><div><Foo/></div></div></div>` looks the same as `<div><Foo/></div>`. The screen (or the web page) is flat so tree hierarchy is not obvious to the end user.

I'm aware that the column layout does not allow the following layout:

```
  +------+------+
  |      |      |
  +------+------+
  |             |
  +-------------+
```

But that layout is very very rare so it's okay for me.

If we can have the flexibility to be able to use the old column behavior, I will be very appreciated.

### Enhancement

I think what's missing here are just two options:

* an option to disable implicit containers
* an option to splitv instead of splith on wide screen by default

Hope this can be approved.

**Comment by Airblader** [comment]
> Say I have the above layout created and before I press Mod+enter, I have no idea what the next state will be

We have the indicator for that which will tell you exactly this.

**Comment by quark-zju** [comment]
@Airblader yes if it's the first window and no if the container already has >= 2 windows.

**Comment by stapelberg** [comment]
I haven’t had time to read the container-related part yet, but:

> an option to splitv instead of splith on wide screen by default

http://i3wm.org/docs/userguide.html#_orientation_for_new_workspaces says:

>  New workspaces get a reasonable default orientation: Wide-screen monitors (anything wider than high) get horizontal orientation, rotated monitors (anything higher than wide) get vertical orientation.
>
> With the default_orientation configuration directive, you can override that behavior.

So, it sounds to me like that option already exists? Does the `default_orientation` option not work for you?

**Comment by quark-zju** [comment]
@stapelberg Yes! Thanks! I vaguely remembered that there is such an option but cannot find it by searching "split" today. So I think we only need an option to disable implicit container, which seems to be easy.

**Comment by stapelberg** [comment]
> To create it, i3 3.x or wmii only needs 5 keystrokes. While i3 4.x requires 2 more (split v), which is 40%, a significant number.

If you frequently create such a layout, I suggest you automate it. That way, you’ll only need a single keystroke (if at all). There are a number of ways to go about automating layout manipulation, but the most promising might be [layout saving/restoring](http://i3wm.org/docs/layout-saving.html).

> Another annoying fact is it is unpredictable. Say I have the above layout created (in an unknown way) and before I press Mod+enter, I have no idea what the next state will be:

It actually is predictable. Here are two screenshots, each one with the default colors and the layout you described:

* created by using [two splitv containers](https://cloud.githubusercontent.com/assets/55506/11622029/23b99f56-9cc2-11e5-87ad-881d856e3494.png)

* created by using [two splith containers](https://cloud.githubusercontent.com/assets/55506/11622030/2741376a-9cc2-11e5-9ed5-8f649aea84bf.png)

As you can see, in the first one, the container on the left side of the screen contains a focused, but inactive window (`client.focused_inactive` in http://i3wm.org/docs/userguide.html#_changing_colors), whereas in the second screenshot, the container on the top side of the screen contains a focused, but inactive window.

I realize it might not be trivial to interpret these clues, but they are there. In addition, if you follow my suggestion and automate the layout creation, you’ll not need to distinguish how the layout was created because it will always be created the same way, allowing you to build up muscle memory.

In the end, we know that the tree model has different trade-offs than a column model, but we chose to go with the tree model nevertheless. I think bottomline it suits our users well, and I’m not intending to change back. Furthermore, I think adding options to restore the old i3 behavior will make the code a nightmare to maintain and has little overall benefit.

I’m sorry I don’t have more positive news for you, but I hope you understand my point of view.

**Comment by quark-zju** [comment]
I am not asking for change back, but a configurable way (even if it is very complex) to use the old behavior. For example, if we have a conditional command like "if a new implicit container is created, then do command xxx", and the option to disable horizontal/vertical implicit container from being created, I can configure myself to have column behavior. It's still the tree layout.

Your screenshots rely on highlighted focused window. You can make the top-left window the inactive focused one, right? Then the two layouts will be indistinguishable. Anyway, I am not arguing on this point. All I wanted is the ability to do column using tree layout.

I create different layouts on demand every day. Layout saving/restore is not interesting to me.

**Comment by stapelberg** [comment]
As I was saying, I don’t think adding an option is a good idea either.

If this is really a big deal for you, the best course of action might be to run a patched version of i3.
